python/cucim/src/cucim/skimage/_shared/distance.py

```diff
@@ -0,0 +1,177 @@
+
+
+__all__ = ["pdist_max_blockwise"]
+
+
+def pdist_max_blockwise(
+    coords,
+    metric="sqeuclidean",
+    *,
+    coords_per_block=4000,
+    compute_argmax=False,
+    cdist_kwargs={},
+):
+    """Find maximum pointwise distance.
+
+    Computes by processing blocks of coordinates to reduce overall memory
+    requirement. The memory used at runtime will be proportional to
+    ``coords_per_block**2``.
+
+    A block size of >= 2000 is recommended to overhead poor GPU resource usage
+    and to reduce kernel launch overhead.
+
+    Parameters
+    ----------
+    coords : np.ndarray (num_points, ndim)
+        The coordinates to process.
+    metric : str, optional
+        Can be any metric supported by `scipy.spatial.distance.cdist`. The
+        default is the squared Euclidean distance (sqeuclidean).
+    coords_per_block : bool, optional
+        Internally, calls to cdist will be made with subsets of coords where
+        the subset size is (coords_per_block, ndim).
+    compute_argmax : bool, optional
+        If True, the value of the coordate indices corresponding to the maxima
+        is returned as the second return Value. Otherwise that value will be
+        ``None``.
+    cdist_kwargs = dict, optional
+        Can provide any additional kwargs to cdist (e.g. `p` for Minkowski
+        norms).
+
+    Returns
+    -------
+    max : float
+        The maximal pairwise distance between the points in `coords`.
+    argmax : 2-tuple of int or None
+        The pair of indices into coords that corresponds to the maxima. If
+        `compute_argmax` is False, this will be None.
+
+    Notes
+    -----
+    The full `cdist` matrix has ``num_coords**2`` elements. Storing this is
+    very wasteful if we  only want to find the maximum pointwise distance. This
+    function processes smaller blocks of the overall cdist result, keeping
+    track of the maximum value (and corrdsponding indices) across all blocks.
+
+    A schematic of the block sizes processed for an array of 10,000 coordinates
+    with a block size of 4000 points is shown below:
+
+      ┌────────┬────────┬────────┐
+      │  4000  │  4000  │  2000  │
+      │  4000  │  4000  │  2000  │
+      ├────────┼────────┼────────┤
+      │   X    │  4000  │  2000  │
+      │        │  4000  │  2000  │
+      ├────────┼────────┼────────┤
+      │   X    │   X    │  2000  │
+      │        │        │  2000  │
+      └────────┴────────┴────────┘
+
+    where due to the symmetry of the pdist matrix, we don't compute the lower
+    triangular blocks.
+    """
+
+    if _distance_on_cpu:
+        warnings.warn(
+            "cuVS >= 25.02 or pylibraft < 24.12 must be installed to use "
+            "GPU-accelerated pairwaise distance computations. Falling back "
+            "to SciPy-based CPU implementation."
+        )
+        xp = np
+        coords = cp.asnumpy(coords)
+    else:
+        xp = cp
+        coords = cp.asarray(coords)
+
+    num_coords, _ = coords.shape
```

Comment:
Input validation might be needed here.

```suggestion

    if not isinstance(coords, (np.ndarray, cp.ndarray)):
        raise TypeError("coords must be a numpy or cupy array")
    
    if coords.ndim != 2:
        raise ValueError(
            f"coords must be a 2-dimensional array, got shape {coords.shape}"
        )
    
    num_coords, _ = coords.shape
```

Reply:
Thanks, it is good to check.

I put this function in the private `_shared` module to indicate it is not part of the public API and only intended for internal use.